onnxruntime/core/framework/provider_bridge_ort.cc

```diff
@@ -49,487 +76,553 @@ struct Provider_IDeviceAllocator_Impl : Provider_IDeviceAllocator {
+  const Provider_OpKernelInfo& Provider_OpKernel_Base__GetInfo(const Provider_OpKernel_Base* p) override { return *reinterpret_cast<const Provider_OpKernelInfo*>(&reinterpret_cast<const OpKernel*>(p)->Info()); }
+
+  // Provider_OpKernelContext
+  const Provider_Tensor* Provider_OpKernelContext__Input_Tensor(const Provider_OpKernelContext* p, int index) override { return reinterpret_cast<const Provider_Tensor*>(reinterpret_cast<const OpKernelContext*>(p)->Input<Tensor>(index)); }
+  Provider_Tensor* Provider_OpKernelContext__Output(Provider_OpKernelContext* p, int index, const TensorShape& shape) override { return reinterpret_cast<Provider_Tensor*>(reinterpret_cast<OpKernelContext*>(p)->Output(index, shape)); }
+
+  // Provider_OpKernelInfo
+  Status Provider_OpKernelInfo__GetAttr_int64(const Provider_OpKernelInfo* p, const std::string& name, int64_t* value) override { return reinterpret_cast<const OpKernelInfo*>(p)->GetAttr(name, value); }
+  Status Provider_OpKernelInfo__GetAttr_float(const Provider_OpKernelInfo* p, const std::string& name, float* value) override { return reinterpret_cast<const OpKernelInfo*>(p)->GetAttr(name, value); }
+
+  const Provider_DataTransferManager& Provider_OpKernelInfo__GetDataTransferManager(const Provider_OpKernelInfo* p) noexcept override { return *reinterpret_cast<const Provider_DataTransferManager*>(&reinterpret_cast<const OpKernelInfo*>(p)->GetDataTransferManager()); }
+  int Provider_OpKernelInfo__GetKernelDef_ExecQueueId(const Provider_OpKernelInfo* p) noexcept override { return reinterpret_cast<const OpKernelInfo*>(p)->GetKernelDef().ExecQueueId(); }
+
+  // Provider_Tensor
+  float* Provider_Tensor__MutableData_float(Provider_Tensor* p) override { return reinterpret_cast<Tensor*>(p)->MutableData<float>(); }
+  const float* Provider_Tensor__Data_float(const Provider_Tensor* p) override { return reinterpret_cast<const Tensor*>(p)->Data<float>(); }
+  const TensorShape& Provider_Tensor__Shape(const Provider_Tensor* p) override { return reinterpret_cast<const Tensor*>(p)->Shape(); }
+
+} provider_host_;
+
+struct ProviderSharedLibrary {
+  ProviderSharedLibrary() {
+    std::string full_path = Env::Default().GetRuntimePath() + std::string(LIBRARY_PREFIX "onnxruntime_providers_shared" LIBRARY_EXTENSION);
+    auto error = Env::Default().LoadDynamicLibrary(full_path, &handle_);
+    if (!error.IsOK()) {
+      LOGS_DEFAULT(ERROR) << error.ErrorMessage();
+      return;
+    }
+
+    void (*PProvider_SetHost)(void*);
+    Env::Default().GetSymbolFromLibrary(handle_, "Provider_SetHost", (void**)&PProvider_SetHost);
+
+    PProvider_SetHost(&provider_host_);
   }
 
-  void LogRuntimeError(uint32_t session_id, const common::Status& status, const char* file, const char* function, uint32_t line) override {
-    return ::onnxruntime::LogRuntimeError(session_id, status, file, function, line);
+  ~ProviderSharedLibrary() {
+    Env::Default().UnloadDynamicLibrary(handle_);
   }
 
-} provider_host_;
+  void* handle_{};
```

Comment:
>void* handle_{} [](start = 2, length = 15)

Prohibit copying. Double destruction would not be good.

Reply:
Done (for both it and ProviderLibrary)